### PR TITLE
Support non-16KB byte arrays for recycling

### PR DESCRIPTION
**Pull request by Tim-Brooks**
The JDK SSLEngine requires that the network packet bytes it operates on
are contiguous byte arrays of a specific length. The default length is
`16709` bytes which is greater than 16KB. This means that we currently
cannot recycle the bytes we use for network buffers. Additionally, we
can only read/write a single packet at a time (without allocating many
buffers). This commit add support for creating a new managed recycler
with a different page size.

### CODE REVIEW

**Comment by tbrooks8** [comment]
I am opening the WIP to see if this a good direction.

The core of this PR is:

1. Add method to the `PageCacheRecycler` that will return a new byte array recycler that is "managed" but the `PageCacheRecycler`. This means that it is closed when the `PageCacheRecycler` is closed. And will use the `MockPageCacheRecycler` release assertions in tests.

2. Adding the `ByteAllocator` class which uses this recycler. This allocator is used by the transport classes to allocate the bytes that they need. If the size requested is larger than the page size, it will just allocate the bytes. This might happen if a large fragment is negotiated during TLS (rare). It will also add a location to add metrics about byte allocation (something we eventually want to add to network metrics).

3. Change the nio transport mechanisms to not depend on the pages being power of two. We could change the page sizes to 32KB, but that means many wasted bytes per page since the ssl engine only requires `16709`.

I can split some of this up as it makes sense. But I just wanted some early feedback on how I created and use new recyclers.

**Review by ywelsch — COMMENTED** [review]
Just some observations I've made while looking at the code of this PoC here. Note that on JDK9, we could also influence the maximum packet size (see `SSLParameters.setMaximumPacketSize`), and keep 16KB byte arrays for recycling.

File: plugins/transport-nio/src/main/java/org/elasticsearch/transport/nio/NioTransport.java [hunk 1]
```diff
@@ -179,4 +191,11 @@ public NioTcpServerChannel createServerChannel(NioSelector selector, ServerSocke
             return nioChannel;
         }
     }
+
+    protected static ByteAllocator createAllocator(Settings settings, PageCacheRecycler pageCacheRecycler, int pageSize) {
+        int numberOfProcessors = EsExecutors.numberOfProcessors(settings);
+        long maxBytes = ByteSizeUnit.MB.toBytes(16) * numberOfProcessors;
```
[the remark this answers is not in the document]
Comment (tbrooks8) [reply plugins/transport-nio/src/main/java/org/elasticsearch/transport/nio/NioTransport.java]: No reason right now. This is just mirror what netty transport will do (16 MB per thread). The amount of bytes dedicated to the nio allocator is something that will probably need to be ironed out over a while.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioTransport.java [hunk 2]
```diff
@@ -67,6 +67,8 @@
 public class SecurityNioTransport extends NioTransport {
     private static final Logger logger = LogManager.getLogger(SecurityNioTransport.class);
 
+    private static final int SSL_PACKET_SIZE = 16709;
```
[the remark this answers is not in the document]
Comment (tbrooks8) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioTransport.java]: There is a constant in java. However it is package private. We could always create an ssl engine and check what it sets the default to. Or we could go with the constant approach above, but have some test verifying that it is the same as the default packet buffer size in the SSLEngine.

**Comment by tbrooks8** [comment]
> Note that on JDK9, we could also influence the maximum packet size (see SSLParameters.setMaximumPacketSize), and keep 16KB byte arrays for recycling.

I'm not sure this is a perfect approach. 

1. It limits us to post-JDK-9.
2. It is changing the default fragment size (TLS fragments are 16KB by default. Now they would need to be smaller).
3. It relies on the other side cooperating using the `max_fragment_extension`. Not everyone has always supported it (https://github.com/openssl/openssl/issues/955).
4. The `max_fragment_extension` has defined values it allows. The next largest value to 16KB is 4KB. Meaning much smaller fragments.
5. The spec only refers to the `max_fragment_length` being negotiated by the client in the CLIENT_HELLO message. Our http server still would need to support the normal sized fragments (16KB + bytes for headers).

Also: FYI I'm not making changes to this PR currently because as I mentioned it is POC and it will probably be split up. It's mainly here to see if there are any objections to the overall approach.

**Comment by jpountz** [comment]
Apologies for the delay, I thought I had replied last week but apparently I had forgotten to hit "Comment".

One thing that is a bit annoying today is the need to have different recyclers for ints, bytes or longs. It would be better to have a single pool to optimize reuse. I'm not a fan of segmenting this even more. Maybe we could increase the page size for the byte[] to 16709 and accept a 2% waste for consumers that require powers of 2 lengths.

**Comment by s1monw** [comment]
just a suggestion, since we will move to java 11 on master maybe we should just change the JDKs default to `16k` for 7x maybe it's fine to use a dedicated pool and a specialized implementation?

**Comment by tbrooks8** [comment]
> just a suggestion, since we will move to java 11 on master maybe we should just change the JDKs default to 16k for 7x maybe it's fine to use a dedicated pool and a specialized implementation?

I think this is based on @ywelsch suggestion and as I mentioned (https://github.com/elastic/elasticsearch/pull/39972#issuecomment-474010549) I do not think this approach will work.

**Comment by tbrooks8** [comment]
This POC is probably not necessary anymore as we have discussed some various options and they will be introduced in other PRs.

**Pull request closed without merging**